### PR TITLE
Add streaming support for JSON

### PR DESCRIPTION
**Pull request by apoorvdeshmukh**
This commit adds support for streaming JSON data to and from Sql Server

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs [hunk 1]
```diff
@@ -1994,7 +1994,7 @@ override public TextReader GetTextReader(int i)
                 {
                     encoding = _metaData[i].encoding;
                 }
-
+                encoding = mt.SqlDbType == SqlDbTypeExtensions.Json ? System.Text.Encoding.UTF8 : encoding;
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs:1997]: ✏️ nit, but I usually find it easier to read if this kind of behavior is expressed as mutually exclusive cases. This way it's clear that there's not something special about the relationship between NCharType and Json that would further change the outcome. Even better would be a switch statement, but then we'd have to unravel the IsNCharType determination.

Something like:
```c#
System.Text.Encoding encoding;
if (mt.SqlDbType == SqlDbTypeExtensions.Json) {
    encoding = System.Text.Encoding.UTF8
}
else if (mt.IsNCharType)
{
    // NChar types always use unicode
    encoding = SqlUnicodeEncoding.SqlUnicodeEncodingInstance;
}
else
{
    encoding = _metaData[i].encoding;
}
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 2]
```diff
@@ -11913,7 +11913,7 @@ private async Task WriteTextFeed(TextDataFeed feed, Encoding encoding, bool need
 
             encoding = encoding ?? TextDataFeed.DefaultEncoding;
 
-            using (ConstrainedTextWriter writer = new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null), encoding), size))
+            using (ConstrainedTextWriter writer = encoding == Encoding.UTF8 ? new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null)), size) : new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null), encoding), size))
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs:11916]: Is this change needed?
These should already be equivalent:
```
new StreamWriter(new TdsOutputStream(this, stateObj, null))
```
```
new StreamWriter(new TdsOutputStream(this, stateObj, null), Encoding.UTF8)
```

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 3]
```diff
@@ -12910,7 +12910,7 @@ private async Task WriteTextFeed(TextDataFeed feed, Encoding encoding, bool need
             char[] inBuff = new char[constTextBufferSize];
 
             encoding = encoding ?? new UnicodeEncoding(false, false);
-            ConstrainedTextWriter writer = new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null), encoding), size);
+            ConstrainedTextWriter writer = encoding == Encoding.UTF8 ? new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null)), size) : new ConstrainedTextWriter(new StreamWriter(new TdsOutputStream(this, stateObj, null), encoding), size);
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs:12913]: See comment above, I don't think this is needed?

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDataReader.cs [hunk 4]
```diff
@@ -2282,7 +2282,7 @@ override public TextReader GetTextReader(int i)
                 {
                     encoding = _metaData[i].encoding;
                 }
-
+                encoding = mt.SqlDbType == SqlDbTypeExtensions.Json ? System.Text.Encoding.UTF8 : encoding;
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDataReader.cs:2285]: See comments above

**Review by mdaigle — CHANGES_REQUESTED** [review]

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 2 above)
Comment (apoorvdeshmukh) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs:11916]: [StreamWriter ](https://learn.microsoft.com/en-us/dotnet/api/system.io.streamwriter.-ctor?view=net-8.0#system-io-streamwriter-ctor(system-io-stream)) without the encoding parameter uses UTF8 encoding with no BOM by default which is what the server also expects for json. In case the encoding is specified explicitly, we need to remove the BOM explicitly before sending the data to the server.
```csharp
        public StreamWriter(Stream stream)
            : this(stream, UTF8NoBOM, DefaultBufferSize, false)
        {
        }

        public StreamWriter(Stream stream, Encoding encoding)
            : this(stream, encoding, DefaultBufferSize, false)
        {
        }
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonStreamTest.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            }
+        }
+
+        private void DeleteFile(string filename)
+        {
+            if (File.Exists(filename))
+            {
+                File.Delete(filename);
+            }
+
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonStreamTest.cs:158]: Please cleanup extra lines

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 2 above)
Comment (mdaigle) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs:11916]: I see. So why not pass in UTF8Encoding.UTF8Encoding() https://learn.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.-ctor?view=net-8.0#system-text-utf8encoding-ctor when we call WriteTextFeed?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs (hunk 1 above)
Comment (apoorvdeshmukh) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs]: Updated with suggested changes.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 2 above)
Comment (apoorvdeshmukh) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: Done

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 3 above)
Comment (apoorvdeshmukh) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs]: Yes. This is not needed anymore.

**Review by mdaigle — APPROVED** [review]

**Review by cheenamalhotra — APPROVED** [review]

**Review by deepaksa1 — APPROVED** [review]
